File: application/src/test/java/com/study/platform/domain/apply/application/ApplyServiceTest.java
package com.study.platform.domain.apply.application;

import com.study.platform.domain.apply.dto.request.ApplyCreateRequest;
import com.study.platform.domain.apply.dto.response.ApplyResponse;
import com.study.platform.domain.apply.event.ApplyApprovedEvent;
import com.study.platform.domain.apply.event.ApplyReceivedEvent;
import com.study.platform.domain.apply.event.ApplyRejectedEvent;
import com.study.platform.domain.apply.model.Apply;
import com.study.platform.domain.apply.model.ApplyStatus;
import com.study.platform.domain.post.model.StudyPost;
import com.study.platform.domain.user.model.User;
import com.study.platform.global.exception.CustomException;
import com.study.platform.global.exception.ErrorCode;
import com.study.platform.support.fake.FakeApplyRepository;
import com.study.platform.support.fake.FakeDomainEventPublisher;
import com.study.platform.support.fake.FakeStudyPostRepository;
import com.study.platform.support.fake.FakeUserRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.test.util.ReflectionTestUtils;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class ApplyServiceTest {

    private FakeApplyRepository applyRepository;
    private FakeStudyPostRepository studyPostRepository;
    private FakeUserRepository userRepository;
    private FakeDomainEventPublisher eventPublisher;
    private ApplyService applyService;

    private UUID authorId;
    private UUID applicantId;
    private UUID postId;
    private User author;
    private User applicant;
    private StudyPost post;

    @BeforeEach
    void setUp() {
        applyRepository = new FakeApplyRepository();
        studyPostRepository = new FakeStudyPostRepository();
        userRepository = new FakeUserRepository();
        eventPublisher = new FakeDomainEventPublisher();
        applyService = new ApplyService(applyRepository, studyPostRepository, userRepository, eventPublisher);

        authorId = UUID.randomUUID();
        applicantId = UUID.randomUUID();
        postId = UUID.randomUUID();

        author = User.create("kakao1", "작성자", "author@test.com");
        ReflectionTestUtils.setField(author, "id", authorId);

        applicant = User.create("kakao2", "지원자", "applicant@test.com");
        ReflectionTestUtils.setField(applicant, "id", applicantId);

        post = StudyPost.create(author, "스터디 모집", "열심히 합니다", "Java", 3, LocalDateTime.now().plusDays(7));
        ReflectionTestUtils.setField(post, "id", postId);

        studyPostRepository.save(post);
        userRepository.save(applicant);
    }

    @Test
    void apply_성공() {
        // given
        ApplyCreateRequest request = new ApplyCreateRequest("지원합니다");

        // when
        ApplyResponse response = applyService.apply(applicantId, postId, request);

        // then
        assertThat(response.status()).isEqualTo(ApplyStatus.PENDING);
        assertThat(applyRepository.existsByPostIdAndApplicantId(postId, applicantId)).isTrue();
        assertThat(eventPublisher.hasEventOf(ApplyReceivedEvent.class)).isTrue();
    }

    @Test
    void apply_게시글없음_예외발생() {
        // given
        ApplyCreateRequest request = new ApplyCreateRequest("지원합니다");

        // when & then
        assertThatThrownBy(() -> applyService.apply(applicantId, UUID.randomUUID(), request))
                .isInstanceOf(CustomException.class)
                .hasFieldOrPropertyWithValue("errorCode", ErrorCode.POST_NOT_FOUND);
    }

    @Test
    void apply_마감된게시글_예외발생() {
        // given
        post.close();
        ApplyCreateRequest request = new ApplyCreateRequest("지원합니다");

        // when & then
        assertThatThrownBy(() -> applyService.apply(applicantId, postId, request))
                .isInstanceOf(CustomException.class)
                .hasFieldOrPropertyWithValue("errorCode", ErrorCode.POST_CLOSED);
    }

    @Test
    void apply_본인게시글지원_예외발생() {
        // given
        ApplyCreateRequest request = new ApplyCreateRequest("지원합니다");

        // when & then
        assertThatThrownBy(() -> applyService.apply(authorId, postId, request))
                .isInstanceOf(CustomException.class)
                .hasFieldOrPropertyWithValue("errorCode", ErrorCode.CANNOT_APPLY_OWN_POST);
    }

    @Test
    void apply_중복지원_예외발생() {
        // given
        Apply existing = Apply.create(post, applicant, "첫 번째 지원", eventPublisher);
        ReflectionTestUtils.setField(existing, "id", UUID.randomUUID());
        applyRepository.save(existing);
        ApplyCreateRequest request = new ApplyCreateRequest("두 번째 지원");

        // when & then
        assertThatThrownBy(() -> applyService.apply(applicantId, postId, request))
                .isInstanceOf(CustomException.class)
                .hasFieldOrPropertyWithValue("errorCode", ErrorCode.ALREADY_APPLIED);
    }

    @Test
    void approve_성공() {
        // given
        Apply apply = Apply.create(post, applicant, "지원합니다", eventPublisher);
        UUID applyId = UUID.randomUUID();
        ReflectionTestUtils.setField(apply, "id", applyId);
        applyRepository.save(apply);

        // when
        ApplyResponse response = applyService.approve(authorId, applyId);

        // then
        assertThat(response.status()).isEqualTo(ApplyStatus.APPROVED);
        assertThat(eventPublisher.hasEventOf(ApplyApprovedEvent.class)).isTrue();
    }

    @Test
    void approve_정원충족시_게시글마감() {
        // given - 이미 승인된 지원 2개 세팅 (maxMembers=3)
        for (int i = 0; i < 2; i++) {
            User u = User.create("kakao" + i, "user" + i, "u" + i + "@test.com");
            ReflectionTestUtils.setField(u, "id", UUID.randomUUID());
            Apply approved = Apply.create(post, u, "지원", eventPublisher);
            ReflectionTestUtils.setField(approved, "id", UUID.randomUUID());
            applyRepository.save(approved);
            approved.approve(eventPublisher);
        }
        Apply apply = Apply.create(post, applicant, "지원합니다", eventPublisher);
        UUID applyId = UUID.randomUUID();
        ReflectionTestUtils.setField(apply, "id", applyId);
        applyRepository.save(apply);

        // when
        applyService.approve(authorId, applyId);

        // then
        assertThat(post.isOpen()).isFalse();
    }

    @Test
    void reject_성공() {
        // given
        Apply apply = Apply.create(post, applicant, "지원합니다", eventPublisher);
        UUID applyId = UUID.randomUUID();
        ReflectionTestUtils.setField(apply, "id", applyId);
        applyRepository.save(apply);

        // when
        ApplyResponse response = applyService.reject(authorId, applyId);

        // then
        assertThat(response.status()).isEqualTo(ApplyStatus.REJECTED);
        assertThat(eventPublisher.hasEventOf(ApplyRejectedEvent.class)).isTrue();
    }

    @Test
    void cancel_성공() {
        // given
        Apply apply = Apply.create(post, applicant, "지원합니다", eventPublisher);
        UUID applyId = UUID.randomUUID();
        ReflectionTestUtils.setField(apply, "id", applyId);
        applyRepository.save(apply);

        // when
        applyService.cancel(applicantId, postId);

        // then
        assertThat(applyRepository.findById(applyId)).isEmpty();
    }

    @Test
    void findApplies_성공() {
        // given
        userRepository.save(author);
        Apply apply = Apply.create(post, applicant, "지원합니다", eventPublisher);
        ReflectionTestUtils.setField(apply, "id", UUID.randomUUID());
        applyRepository.save(apply);

        // when
        List<ApplyResponse> responses = applyService.findApplies(authorId, postId);

        // then
        assertThat(responses).hasSize(1);
    }

    @Test
    void findApplies_작성자아님_예외발생() {
        // given
        UUID otherId = UUID.randomUUID();

        // when & then
        assertThatThrownBy(() -> applyService.findApplies(otherId, postId))
                .isInstanceOf(CustomException.class)
                .hasFieldOrPropertyWithValue("errorCode", ErrorCode.FORBIDDEN);
    }
}

File: application/src/test/kotlin/com/study/platform/domain/apply/application/ApplyServiceTest.kt
package com.study.platform.domain.apply.application

import com.study.platform.domain.apply.dto.request.ApplyCreateRequest
import com.study.platform.domain.apply.dto.response.ApplyResponse
import com.study.platform.domain.apply.event.ApplyApprovedEvent
import com.study.platform.domain.apply.event.ApplyReceivedEvent
import com.study.platform.domain.apply.event.ApplyRejectedEvent
import com.study.platform.domain.apply.model.ApplyStatus
import com.study.platform.domain.post.model.StudyPost
import com.study.platform.domain.user.model.User
import com.study.platform.global.exception.CustomException
import com.study.platform.global.exception.ErrorCode
import com.study.platform.support.TestFixtures
import com.study.platform.support.fake.FakeApplyQueryPort
import com.study.platform.support.fake.FakeApplyRepository
import com.study.platform.support.fake.FakeDomainEventPublisher
import com.study.platform.support.fake.FakeStudyPostRepository
import com.study.platform.support.fake.FakeUserRepository
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.data.domain.PageRequest
import java.util.UUID

class ApplyServiceTest {

    private lateinit var applyRepository: FakeApplyRepository
    private lateinit var studyPostRepository: FakeStudyPostRepository
    private lateinit var userRepository: FakeUserRepository
    private lateinit var eventPublisher: FakeDomainEventPublisher

    private lateinit var findAppliesService: FindAppliesService
    private lateinit var applyStudyPostService: ApplyStudyPostService
    private lateinit var cancelApplyService: CancelApplyService
    private lateinit var approveApplyService: ApproveApplyService
    private lateinit var rejectApplyService: RejectApplyService

    private lateinit var authorId: UUID
    private lateinit var applicantId: UUID
    private lateinit var postId: UUID
    private lateinit var author: User
    private lateinit var applicant: User
    private lateinit var post: StudyPost

    @BeforeEach
    fun setUp() {
        applyRepository = FakeApplyRepository()
        studyPostRepository = FakeStudyPostRepository()
        userRepository = FakeUserRepository()
        eventPublisher = FakeDomainEventPublisher()

        findAppliesService = FindAppliesService(FakeApplyQueryPort(applyRepository), studyPostRepository)
        applyStudyPostService = ApplyStudyPostService(applyRepository, studyPostRepository, userRepository, eventPublisher)
        cancelApplyService = CancelApplyService(applyRepository)
        approveApplyService = ApproveApplyService(applyRepository, studyPostRepository, eventPublisher)
        rejectApplyService = RejectApplyService(applyRepository, eventPublisher)

        authorId = UUID.randomUUID()
        applicantId = UUID.randomUUID()
        postId = UUID.randomUUID()

        author = TestFixtures.createUser(id = authorId, kakaoId = "kakao1", nickname = "작성자", email = "author@test.com")
        applicant = TestFixtures.createUser(id = applicantId, kakaoId = "kakao2", nickname = "지원자", email = "applicant@test.com")
        post = TestFixtures.createStudyPost(id = postId, author = author, techStack = "Java", maxMembers = 3)

        studyPostRepository.save(post)
        userRepository.save(applicant)
    }

    @Test
    fun `apply_성공`() {
        // given
        val request = ApplyCreateRequest("지원합니다")

        // when
        val response: ApplyResponse = applyStudyPostService.execute(applicantId, postId, request)

        // then
        assertThat(response.status).isEqualTo(ApplyStatus.PENDING)
        assertThat(applyRepository.existsByPostIdAndApplicantId(postId, applicantId)).isTrue()
        assertThat(eventPublisher.hasEventOf(ApplyReceivedEvent::class.java)).isTrue()
    }

    @Test
    fun `apply_게시글없음_예외발생`() {
        // given
        val request = ApplyCreateRequest("지원합니다")

        // when & then
        assertThatThrownBy { applyStudyPostService.execute(applicantId, UUID.randomUUID(), request) }
            .isInstanceOf(CustomException::class.java)
            .hasFieldOrPropertyWithValue("errorCode", ErrorCode.POST_NOT_FOUND)
    }

    @Test
    fun `apply_마감된게시글_예외발생`() {
        // given
        post.close()
        val request = ApplyCreateRequest("지원합니다")

        // when & then
        assertThatThrownBy { applyStudyPostService.execute(applicantId, postId, request) }
            .isInstanceOf(CustomException::class.java)
            .hasFieldOrPropertyWithValue("errorCode", ErrorCode.POST_CLOSED)
    }

    @Test
    fun `apply_본인게시글지원_예외발생`() {
        // given
        val request = ApplyCreateRequest("지원합니다")

        // when & then
        assertThatThrownBy { applyStudyPostService.execute(authorId, postId, request) }
            .isInstanceOf(CustomException::class.java)
            .hasFieldOrPropertyWithValue("errorCode", ErrorCode.CANNOT_APPLY_OWN_POST)
    }

    @Test
    fun `apply_중복지원_예외발생`() {
        // given
        val existing = TestFixtures.createApply(post = post, applicant = applicant, message = "첫 번째 지원")
        applyRepository.save(existing)
        val request = ApplyCreateRequest("두 번째 지원")

        // when & then
        assertThatThrownBy { applyStudyPostService.execute(applicantId, postId, request) }
            .isInstanceOf(CustomException::class.java)
            .hasFieldOrPropertyWithValue("errorCode", ErrorCode.ALREADY_APPLIED)
    }

    @Test
    fun `approve_성공`() {
        // given
        val applyId = UUID.randomUUID()
        val apply = TestFixtures.createApply(id = applyId, post = post, applicant = applicant)
        applyRepository.save(apply)

        // when
        val response: ApplyResponse = approveApplyService.execute(authorId, applyId)

        // then
        assertThat(response.status).isEqualTo(ApplyStatus.APPROVED)
        assertThat(eventPublisher.hasEventOf(ApplyApprovedEvent::class.java)).isTrue()
    }

    @Test
    fun `approve_정원충족시_게시글마감`() {
        // given
        for (i in 0 until 2) {
            val u = TestFixtures.createUser(kakaoId = "kakao$i", nickname = "user$i", email = "u$i@test.com")
            val approved = TestFixtures.createApply(post = post, applicant = u, message = "지원")
            applyRepository.save(approved)
            approved.approve(eventPublisher)
        }
        val applyId = UUID.randomUUID()
        val apply = TestFixtures.createApply(id = applyId, post = post, applicant = applicant)
        applyRepository.save(apply)

        // when
        approveApplyService.execute(authorId, applyId)

        // then
        assertThat(post.isOpen()).isFalse()
    }

    @Test
    fun `reject_성공`() {
        // given
        val applyId = UUID.randomUUID()
        val apply = TestFixtures.createApply(id = applyId, post = post, applicant = applicant)
        applyRepository.save(apply)

        // when
        val response: ApplyResponse = rejectApplyService.execute(authorId, applyId)

        // then
        assertThat(response.status).isEqualTo(ApplyStatus.REJECTED)
        assertThat(eventPublisher.hasEventOf(ApplyRejectedEvent::class.java)).isTrue()
    }

    @Test
    fun `cancel_성공`() {
        // given
        val applyId = UUID.randomUUID()
        val apply = TestFixtures.createApply(id = applyId, post = post, applicant = applicant)
        applyRepository.save(apply)

        // when
        cancelApplyService.execute(applicantId, postId)

        // then
        assertThat(applyRepository.findById(applyId)).isNull()
    }

    @Test
    fun `findApplies_성공`() {
        // given
        userRepository.save(author)
        val apply = TestFixtures.createApply(post = post, applicant = applicant)
        applyRepository.save(apply)

        // when
        val responses = findAppliesService.execute(authorId, postId, PageRequest.of(0, 20))

        // then
        assertThat(responses.content).hasSize(1)
    }

    @Test
    fun `findApplies_작성자아님_예외발생`() {
        // given
        val otherId = UUID.randomUUID()

        // when & then
        assertThatThrownBy { findAppliesService.execute(otherId, postId, PageRequest.of(0, 20)) }
            .isInstanceOf(CustomException::class.java)
            .hasFieldOrPropertyWithValue("errorCode", ErrorCode.FORBIDDEN)
    }
}
